crates/escrow/src/lib.rs

```diff
@@ -0,0 +1,602 @@
+//! # Escrow Module
+//! Receive vote-escrowed tokens for locking the native currency.
+
+#![deny(warnings)]
+#![cfg_attr(test, feature(proc_macro_hygiene))]
+#![cfg_attr(not(feature = "std"), no_std)]
+
+#[cfg(any(feature = "runtime-benchmarks", test))]
+mod benchmarking;
+
+mod default_weights;
+pub use default_weights::WeightInfo;
+
+#[cfg(test)]
+mod mock;
+
+#[cfg(test)]
+mod tests;
+
+use codec::{Decode, Encode};
+use frame_support::{
+    ensure,
+    traits::{
+        BalanceStatus, Currency, ExistenceRequirement, Get, LockIdentifier, LockableCurrency, ReservableCurrency,
+        SignedImbalance, WithdrawReasons,
+    },
+    transactional,
+};
+use scale_info::TypeInfo;
+use sp_runtime::{
+    traits::{AtLeast32BitUnsigned, Convert, Saturating, Zero},
+    DispatchError, DispatchResult,
+};
+use sp_std::marker::PhantomData;
+
+const LOCK_ID: LockIdentifier = *b"escrowed";
+
+type BalanceOf<T> = <<T as Config>::Currency as Currency<<T as frame_system::Config>::AccountId>>::Balance;
+type PositiveImbalanceOf<T> =
+    <<T as Config>::Currency as Currency<<T as frame_system::Config>::AccountId>>::PositiveImbalance;
+type NegativeImbalanceOf<T> =
+    <<T as Config>::Currency as Currency<<T as frame_system::Config>::AccountId>>::NegativeImbalance;
+type MaxLocksOf<T> = <<T as Config>::Currency as LockableCurrency<<T as frame_system::Config>::AccountId>>::MaxLocks;
+
+#[derive(Default, Encode, Decode, Debug, Clone, TypeInfo)]
+pub struct Point<Balance, BlockNumber> {
+    bias: Balance,
+    slope: Balance,
+    ts: BlockNumber,
+}
+
+impl<Balance: AtLeast32BitUnsigned + Copy, BlockNumber: AtLeast32BitUnsigned + Copy> Point<Balance, BlockNumber> {
+    fn new<BlockNumberToBalance: Convert<BlockNumber, Balance>>(
+        amount: Balance,
+        start_height: BlockNumber,
+        end_height: BlockNumber,
+        max_period: BlockNumber,
+    ) -> Self {
+        let max_period = BlockNumberToBalance::convert(max_period);
+        let height_diff = BlockNumberToBalance::convert(end_height - start_height);
+
+        let slope = amount / max_period;
+        let bias = slope * height_diff;
+
+        Self {
+            bias,
+            slope,
+            ts: start_height,
+        }
+    }
+
+    fn balance_at<BlockNumberToBalance: Convert<BlockNumber, Balance>>(&self, height: BlockNumber) -> Balance {
+        let height_diff = BlockNumberToBalance::convert(height - self.ts);
```

Comment:
Isn't `self.ts` in the future and thus `self.ts > height` -> doesn't this overflow?

Reply:
No `self.ts` is likely always in the past, but I will adjust the implementation to use saturated math here.